Create SSTI-prevention.md (#207)
cobalthq/cobalt-product-public-docs: content/en/BestPractices/SSTI-prevention.md

---

> title: Protect against Server-Side Template Injection
> linkTitle: Protect against SSTI
> weight: 620

Reviewer: ```suggestion
title: Prevent Server-Side Template Injections
linkTitle: Prevent SSTI
weight: 670
```

description: >
  Keep your app safe from SSTI attacks.
---

{{% pageinfo %}}

> Learn about Server-Side Template Injection (SSTI) exploitation and discover how to prevent them.

Reviewer: ```suggestion
Learn about the Server-Side Template Injection (SSTI) vulnerability, and discover how to prevent it.
```

{{% /pageinfo %}}

> 
> 
> SSTI is categorised under injection in the [OWASP Top 10 Web Application Security Risks (2021)](https://owasp.org/www-project-top-ten/) and is ranked 3rd according to its impact and likelihood.
> 
> 
> ## Description
> 
> The server-side template injection vulnerability allows an attacker to execute server-side commands by injecting malicious data into a template. A vulnerability occurs when invalid user input is embedded in a template engine. This can result in [remote code execution (RCE)](https://www.imperva.com/learn/application-security/remote-code-execution/).
> 
> For example, depending on the IP that accesses a site, the site may look like this:

Reviewer: ```suggestion
OWASP includes SSTI in the [Top 10 Web Application Security Risks (2021)](https://owasp.org/www-project-top-ten/) and lists it under [Injection](https://owasp.org/Top10/A03_2021-Injection/).

## Description

A server-side template injection allows an attacker to execute server-side commands by injecting malicious data into a template. A vulnerability occurs when a threat actor embeds invalid user input in a template engine. This can result in [remote code execution (RCE)](https://www.imperva.com/learn/application-security/remote-code-execution/).

For example, depending on the IP address from which a user accesses a site, an injection may look like this:
```

> ```ts
> <h1>Welcome to the cobalt!</h1>
> <u>This page is being accessed from the remote address: {{ip}}</u>
> ```

Reviewer: ```suggestion
```html
<h1>Welcome to Example Project!</h1>
<u>You're accessing this page from a remote address: {{ip}}</u>
```
```

> 
> 
> SSTI is categorised under injection in the [OWASP Top 10 Web Application Security Risks (2021)](https://owasp.org/www-project-top-ten/) and is ranked 3rd according to its impact and likelihood.
> 
> 
> ## Description
> 
> The server-side template injection vulnerability allows an attacker to execute server-side commands by injecting malicious data into a template. A vulnerability occurs when invalid user input is embedded in a template engine. This can result in [remote code execution (RCE)](https://www.imperva.com/learn/application-security/remote-code-execution/).
> 
> For example, depending on the IP that accesses a site, the site may look like this:
> 
> ```ts
> <h1>Welcome to the cobalt!</h1>
> <u>This page is being accessed from the remote address: {{ip}}</u>
> ```
> 
> It will render the remote address into the {{ip}} variable while reusing the rest of the HTML for every request it receives.
> 
> Typically, server-side template injection has a devastating effect. It generally results in remote code execution by completely controlling the back-end server when a server-side template injection vulnerability is exploited. Despite not executing any code, the attacker may still be able to access sensitive server data.

Reviewer: ```suggestion
OWASP includes SSTI in the [Top 10 Web Application Security Risks (2021)](https://owasp.org/www-project-top-ten/) and lists it under [Injection](https://owasp.org/Top10/A03_2021-Injection/).

## Description

A server-side template injection allows an attacker to execute server-side commands by injecting malicious data into a template. A vulnerability occurs when a threat actor embeds invalid user input in a template engine. This can result in [remote code execution (RCE)](https://www.imperva.com/learn/application-security/remote-code-execution/).

For example, depending on the IP address from which a user accesses a site, an injection may look like this:

```html
<h1>Welcome to Example Project!</h1>
<u>You're accessing this page from a remote address: {{ip}}</u>
```

The injection passes a remote address into the `{{ip}}` variable while reusing the rest of the HTML code for every request it receives.

SSTI may have a devastating effect. It generally results in remote code execution by controlling the backend server. Even if the attacker doesn't execute any code, they may still access sensitive server data.
```

> ```ts
> <h1>Welcome to the cobalt!</h1>
> <u>This page is being accessed from the remote address: {{ip}}</u>
> ```
> 
> It will render the remote address into the {{ip}} variable while reusing the rest of the HTML for every request it receives.
> 
> Typically, server-side template injection has a devastating effect. It generally results in remote code execution by completely controlling the back-end server when a server-side template injection vulnerability is exploited. Despite not executing any code, the attacker may still be able to access sensitive server data.

Reviewer: ```suggestion
```html
<h1>Welcome to Example Project!</h1>
<u>You're accessing this page from a remote address: {{ip}}</u>
```

The injection passes a remote address into the `{{ip}}` variable while reusing the rest of the HTML code for every request it receives.

SSTI may have a devastating effect. It generally results in remote code execution by controlling the backend server. Even if the attacker doesn't execute any code, they may still access sensitive server data.
```

> It will render the remote address into the {{ip}} variable while reusing the rest of the HTML for every request it receives.
> 
> Typically, server-side template injection has a devastating effect. It generally results in remote code execution by completely controlling the back-end server when a server-side template injection vulnerability is exploited. Despite not executing any code, the attacker may still be able to access sensitive server data.

Reviewer: ```suggestion
The injection passes a remote address into the `{{ip}}` variable while reusing the rest of the HTML code for every request it receives.

SSTI may have a devastating effect. It generally results in remote code execution by controlling the backend server. Even if the attacker doesn't execute any code, they may still access sensitive server data.
```

## Attack Scenarios

> 
> ### Twig template
> 
> 
> This code example illustrates SSTI vulnerability in the `getFilter()` function that uses `call_user_func($callback, $name)`. From the HTTP GET request, the server retrieves and executes the `name` parameter:
> 

Reviewer: ```suggestion
Let's see some example attack scenarios.

### Twig Template

This code example illustrates an SSTI vulnerability in the `getFilter()` function that uses `call_user_func($callback, $name)`. From the HTTP GET request, the server retrieves and executes the `name` parameter:
```

> ```ts

Reviewer: Is it a PHP code block?

Author: Hello. @ana-dashuk-cobalt  Yes, the twig is the PHP template engine.

> ```ts

Reviewer: ```suggestion
```php
```

public function getFilter($name)
{
        [snip]
        foreach ($this->filterCallbacks as $callback) {
        if (false !== $filter = call_user_func($callback, $name)) {
            return $filter;
        }
    }
    return false;
}

$output = $twig->render($_GET['name'],  array("first_name" => $user.first_name) );
```

#### Analysis

> In the above example, the user controls the template's content via the `name` GET parameter rather than a value passed into it.

Reviewer: ```suggestion
In this example, the user controls the template content using the `name` GET parameter rather than a value passed into it.
```

#### Impact

> An arbitrary shell command can be executed easily by registering `exec` as a filter callback and then calling `getFilter()`.
> 
> An attacker can execute system commands using the "env" attribute of the "Twig_Environment" object.
> 
> 

Reviewer: ```suggestion
An attacker can execute an arbitrary shell command by registering `exec` as a filter callback and then calling `getFilter()`.

They can use the `env` attribute of the `Twig_Environment` object to run system commands.
```

> ```ts
> Attacker’s input via name parameter,
> 
> {{_self.env.registerUndefinedFilterCallback("exec")}}{{_self.env.getFilter("id")}}
> ```

Reviewer: ```suggestion
```ts
{{_self.env.registerUndefinedFilterCallback("exec")}}{{_self.env.getFilter("id")}}
```
```

#### Prevention

> Everything works fine if the name is passed into the template, as in the following example.

Reviewer: ```suggestion
To prevent this, pass a name into the template, as in the following example:
```

```ts
$output = $twig->render("Dear {name},", array("first_name" => $user.first_name) ); 
```

> 
> 
> 
> ### Jinja2
> 
> A vulnerability exists in the following Jinja2 code because the user-supplied variable `name` is concatenated to the template string. When an HTTP GET request is made, the page function accepts a `name` parameter and renders an HTML response with the contents of the name variable.
> 
> 

Reviewer: ```suggestion
### Jinja2

This Jinja2 code contains a vulnerability. The variable `name` that a user passes is concatenated to the template string.
```

```ts
@app.route("/page")
def page():
    name = request.values.get('name')
    output = Jinja2.from_string('Hello ' + name + '!').render()
    return output
```

> 
> 
> #### Analysis
> 
> When an HTTP GET request is made, the page function accepts a 'name' parameter and renders an HTML response with the contents of the name variable.
> 
> #### Impact
> 
> As a result of this usage of templates, both XSS vulnerabilities and SSTI vulnerabilities are created. The following code block is an example of an attacker’s input via the `name` parameter.

Reviewer: ```suggestion

#### Analysis

When you send an HTTP GET request, the page function accepts the `name` parameter and renders an HTML response with the content of the variable.

#### Impact

Improper use of templates may lead to both [Cross-Site Scripting (XSS)](https://owasp.org/www-community/attacks/xss/) and SSTI vulnerabilities. The following code block is an example of a malicious input through the `name` parameter.
```

Reviewer: @unstabl3 For consistency, can we also add a Prevention section here?

Author: @ana-dashuk-cobalt, the prevention for `jinja2` is already added. The same prevention is needed for regular and `bypassing the blacklist`.

```ts

> {{config.__class__.__init__.__globals__['os'].popen('ls').read()}}

Reviewer: @unstabl3 Can you please double-check this code sample? Maybe I don't have a full context (and my knowledge is limited), but I don't see the `name` parameter here.

Author: @ana-dashuk-cobalt we don't need to give a `name` parameter in the code as we have already mentioned above that the following code will be submitted via URL.

```

> 
> 
> ### Bypassing the blacklisting
> 
> The problem with blacklists is that they can often be circumvented, which is terrible.

Reviewer: ```suggestion

### Bypassing Blacklists

Attackers may circumvent the blacklists that you configured.
```

 
```ts
@app.route("/page")
blacklist = ["__class__",”request[request.”]
def page():
    name = request.values.get('name')
    for bad_string in blacklist:
        if  bad_string in name:
            return "HACKING ATTEMPT {}".format(bad_string), 400
        else:
            output = Jinja2.from_string('Hello ' + name + '!').render()
            return output    
```

#### Analysis

> It is possible to retrieve the value of a new `cobalt` GET parameter using the `request.args.param`. An attacker can bypass `__class__` using this. Since we have also blacklisted `request.args.param`, we will use jinja2 native function called `| attr().` 
> 

Reviewer: ```suggestion
A threat actor can bypass `__class__` and retrieve the value of a new GET parameter using `request.args.param`. Because we included `request.args.param` in the blacklist, we can use a native Jinja2 function `| attr().`
```

#### Impact

An attacker can pass the following input to the server using the `name` parameter.

```ts
name={{request|attr('application')|attr('\x5f\x5fglobals\x5f\x5f')|attr('\x5f\x5fgetitem\x5f\x5f')('\x5f\x5fbuiltins\x5f\x5f')|attr('\x5f\x5fgetitem\x5f\x5f')('\x5f\x5fimport\x5f\x5f')('os')|attr('popen')('id')|attr('read')()}}
```

#### Prevention

> The variable `name` should be passed to the template context.

Reviewer: ```suggestion
Pass the `name` variable to the template context.
```

> The variable `name` should be passed to the template context.
> 

Reviewer: ```suggestion
Pass the `name` variable to the template context.
```

```ts
@app.route("/page")
def page():
    name = request.values.get('name')
    output = Jinja2.from_string('Hello {{name}}!').render(name = name)
    return output
```

> 
> ## Best practices
> 
> A template engine has advantages and facilitates modifications. For example, a design change can be made without touching the code flow and autonomously by the designers/integrators. The changes are independent of the code and logical processing. In the same way, developers can change logical elements and code without affecting data display or breaking the interface. It makes maintaining the project easier.
> For SSTI prevention, never allow users to modify or create templates. However, when business needs arise, there are solutions:

Reviewer: ```suggestion
## Best Practices

A template engine allows you to streamline the process of updating and maintaining your application.

For example, designers or integrators can autonomously make a design change without touching the code flow. Their changes are independent of the code and logical processing. Similarly, developers can change logical elements and code without affecting data display or breaking the interface.

To prevent an SSTI attack, never allow users to modify or create templates. If you have no other choice, follow our recommendations.
```

> #### Sanitization
> Before using a template, detecting and removing potentially malicious content is necessary. Data transmitted by users must therefore be efficiently analyzed. Various methods can be used to accomplish this (e.g., regex, whitelists of authorised expressions).
> There is, however, a drawback to this solution. In addition, configuration errors can put your environment at risk.

Reviewer: ```suggestion
### Sanitize Data

Before using a template, identify and remove potentially malicious content.

Make sure to efficiently analyze data that users pass. You can use various methods to do this, such as regular expression tools or whitelists of authorized expressions.

However, there is a drawback to this solution. Configuration errors can put your environment at risk.
```

> #### Sandboxing
> The aim is to provide a closed environment where risky modules and features are disabled. For example, if user-supplied data is interpreted, it cannot provide access to other components or data.
> It may be challenging to create a sandbox environment, and misconfigurations may result in bypassing the environment.

Reviewer: ```suggestion
### Set Up a Sandbox Environment

You can turn off risky modules and features in a closed sandbox environment. This allows you to limit access to other application components.

It may be challenging to create a sandbox environment. Misconfigurations may result in bypassing the environment.
```

> #### Logic less template
> A logic-less engine template exists, which separates visual rendering from code interpretation. Mustache is one of the most popular.
> In Mustache templates, none of the control flow is explicit, as data drive all control. As a result, application logic cannot be integrated into Mustache templates. It is best to keep the logic and presentation as separate as possible to reduce the risk of template-based attacks.
> Therefore the risk of RCE attacks has therefore been significantly reduced.

Reviewer: ```suggestion
### Use Logic-Less Templates

A logic-less template engine separates visual rendering from code interpretation. [Mustache](https://github.com/mustache/mustache) is one of the most popular engines.

In Mustache templates, none of the control flows is explicit because data drives all the control. As a result, you can't integrate your application logic into templates.

Keep the application logic and presentation as separate as possible to reduce the risk of template-based attacks. This allows you to prevent remote code execution (RCE) attacks.
```
